Seems something went weird with docker in jenkins: 

```
16:55:49 Waiting for MySQL(utf8mb4) initialisation ...
17:05:50 ................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
17:05:50 [ERROR] Waited 600 seconds, no response
17:05:50 Kill the docker adab554079d8ef09d2081be106de042dce91b7e6e06734078fa2513676902a47
17:05:50 adab554079d8ef09d2081be106de042dce91b7e6e06734078fa2513676902a47
17:05:50 adab554079d8ef09d2081be106de042dce91b7e6e06734078fa2513676902a47
17:05:50 Makefile:179: recipe for target 'test-php' failed
17:05:50 make: *** [test-php] Error 1
[Pipeline] }
[Pipeline] // timeout
[Pipeline] echo
17:05:50 Test execution failed: hudson.AbortException: script returned exit code 2
[Pipeline] step
17:05:50 Recording test results```

--- a/lib/private/Repair/Apps.php
+++ b/lib/private/Repair/Apps.php
@@ -69,6 +69,10 @@ class Apps implements IRepairStep {
 		return 'Upgrade app code from the marketplace';
 	}
 
+	/**
+	 * Are we updating from an older version?
+	 * @return bool
+	 */
 	private function isCoreUpdate() {
 		$installedVersion = $this->config->getSystemValue('version', '0.0.0');
 		$currentVersion = implode('.', \OCP\Util::getVersion());
@@ -86,35 +90,54 @@ class Apps implements IRepairStep {
 	public function run(IOutput $output) {
 		$isCoreUpdate = $this->isCoreUpdate();
 		$appsToUpgrade = $this->getAppsToUpgrade();
-		$isMarketEnabled = $this->appManager->isEnabledForUser('market');
 		$failedCompatibleApps = [];
 		$failedMissingApps = $appsToUpgrade[self::KEY_MISSING];
 		$failedIncompatibleApps = $appsToUpgrade[self::KEY_INCOMPATIBLE];
 		$hasNotUpdatedCompatibleApps = 0;
 
-		if ($isMarketEnabled && $isCoreUpdate) {
-			$this->loadApp('market');
-			try {
+		if($isCoreUpdate) {
+			// Then we need to enable the market app to support app updates / downloads during upgrade
+			$output->info('Enabling market app to assist with update');
+			$this->appManager->enableApp('market');
+		}
+
+		// Use market to fix missing / old apps
+		$this->loadApp('market');
+		$output->info('Using market to update existing apps');
+		try {
+			// Try to update incompatible apps
+			if($appsToUpgrade[self::KEY_INCOMPATIBLE]) {
+				$output->info('Attempting to update the following existing but incompatible app from market: '.implode(', ', $appsToUpgrade[self::KEY_INCOMPATIBLE]));
 				$failedIncompatibleApps = $this->getAppsFromMarket(
 					$output,
 					$appsToUpgrade[self::KEY_INCOMPATIBLE],
 					'upgradeAppStoreApp'
 				);
+			}
+
+			// Try to download missing apps
+			if($appsToUpgrade[self::KEY_MISSING]) {
+				$output->info('Attempting to update the following missing apps from market: '.implode(', ', $appsToUpgrade[self::KEY_MISSING]));
 				$failedMissingApps = $this->getAppsFromMarket(
 					$output,
 					$appsToUpgrade[self::KEY_MISSING],
 					'reinstallAppStoreApp'
 				);
+			}
+
+			// Try to update compatible apps
+			if($appsToUpgrade[self::KEY_COMPATIBLE]) {
+				$output->info('Attempting to update the following existing compatible apps from market: '.implode(', ', $appsToUpgrade[self::KEY_MISSING]));
 				$failedCompatibleApps = $this->getAppsFromMarket(
 					$output,
 					$appsToUpgrade[self::KEY_COMPATIBLE],
 					'upgradeAppStoreApp'
 				);
-				$hasNotUpdatedCompatibleApps = count($failedCompatibleApps);
-			} catch (AppManagerException $e) {
-				$output->warning('No connection to marketplace: ' . $e->getPrevious());
-				$isMarketEnabled = false;
 			}
+
+			$hasNotUpdatedCompatibleApps = count($failedCompatibleApps);
+		} catch (AppManagerException $e) {
+			$output->warning('No connection to marketplace: ' . $e->getPrevious());
 		}
 
 		$hasBlockingMissingApps = count($failedMissingApps);
@@ -122,9 +145,9 @@ class Apps implements IRepairStep {
 
 		if ($hasBlockingIncompatibleApps || $hasBlockingMissingApps) {
 			// fail
-			$output->warning('You have incompatible or missing apps enabled.');
+			$output->warning('You have incompatible or missing apps enabled that could not be found or updated via the marketplace.');
 			$output->warning(
-				'please install app manually with tarball or disable them'
+				'please install app manually with tarball or disable them with:'
 				. $this->getOccDisableMessage(array_merge($failedIncompatibleApps, $failedMissingApps))
 			);
 

--- a/lib/private/Repair/Apps.php
+++ b/lib/private/Repair/Apps.php
@@ -179,12 +179,16 @@ class Apps implements IRepairStep {
 					new GenericEvent($app)
 				);
 			} catch (AppAlreadyInstalledException $e) {
+				$output->info($e->getMessage());
 				$failedApps[] = $app;
 			} catch (AppNotInstalledException $e) {
+				$output->info($e->getMessage());
 				$failedApps[] = $app;
 			} catch (AppNotFoundException $e) {
+				$output->info($e->getMessage());
 				$failedApps[] = $app;
 			} catch (AppUpdateNotFoundException $e) {
+				$output->info($e->getMessage());
 				$failedApps[] = $app;
 			} catch (AppManagerException $e) {
 				// No connection to market. Abort.

--- a/lib/private/Repair/Apps.php
+++ b/lib/private/Repair/Apps.php
@@ -84,6 +84,19 @@ class Apps implements IRepairStep {
 	}
 
 	/**
+	 * If we are updating from <= 10.0.0 we need to enable the marketplace before running the update
+	 * @return bool
+	 */
+	private function requiresMarketEnable() {
+		$installedVersion = $this->config->getSystemValue('version', '0.0.0');
+		$versionDiff = version_compare('10.0.0', $installedVersion);
+		if ($versionDiff >= 0) {
+			return true;
+		}
+		return false;
+	}
+
+	/**
 	 * @param IOutput $output
 	 * @throws RepairException
 	 */
@@ -94,50 +107,57 @@ class Apps implements IRepairStep {
 		$failedMissingApps = $appsToUpgrade[self::KEY_MISSING];
 		$failedIncompatibleApps = $appsToUpgrade[self::KEY_INCOMPATIBLE];
 		$hasNotUpdatedCompatibleApps = 0;
+		$requiresMarketEnable = $this->requiresMarketEnable();
 
-		if($isCoreUpdate) {
+		if($isCoreUpdate && $requiresMarketEnable) {
 			// Then we need to enable the market app to support app updates / downloads during upgrade
 			$output->info('Enabling market app to assist with update');
 			$this->appManager->enableApp('market');
 		}
 
-		// Use market to fix missing / old apps
-		$this->loadApp('market');
-		$output->info('Using market to update existing apps');
-		try {
-			// Try to update incompatible apps
-			if($appsToUpgrade[self::KEY_INCOMPATIBLE]) {
-				$output->info('Attempting to update the following existing but incompatible app from market: '.implode(', ', $appsToUpgrade[self::KEY_INCOMPATIBLE]));
-				$failedIncompatibleApps = $this->getAppsFromMarket(
-					$output,
-					$appsToUpgrade[self::KEY_INCOMPATIBLE],
-					'upgradeAppStoreApp'
-				);
-			}
+		// Check if we can use the marketplace to update apps as needed?
+		if($this->appManager->isEnabledForUser('market')) {
+			// Use market to fix missing / old apps
+			$this->loadApp('market');
+			$output->info('Using market to update existing apps');
+			try {
+				// Try to update incompatible apps
+				if(!empty($appsToUpgrade[self::KEY_INCOMPATIBLE])) {
+					$output->info('Attempting to update the following existing but incompatible app from market: '.implode(', ', $appsToUpgrade[self::KEY_INCOMPATIBLE]));
+					$failedIncompatibleApps = $this->getAppsFromMarket(
+						$output,
+						$appsToUpgrade[self::KEY_INCOMPATIBLE],
+						'upgradeAppStoreApp'
+					);
+				}
 
-			// Try to download missing apps
-			if($appsToUpgrade[self::KEY_MISSING]) {
-				$output->info('Attempting to update the following missing apps from market: '.implode(', ', $appsToUpgrade[self::KEY_MISSING]));
-				$failedMissingApps = $this->getAppsFromMarket(
-					$output,
-					$appsToUpgrade[self::KEY_MISSING],
-					'reinstallAppStoreApp'
-				);
-			}
+				// Try to download missing apps
+				if(!empty($appsToUpgrade[self::KEY_MISSING])) {
+					$output->info('Attempting to update the following missing apps from market: '.implode(', ', $appsToUpgrade[self::KEY_MISSING]));
+					$failedMissingApps = $this->getAppsFromMarket(
+						$output,
+						$appsToUpgrade[self::KEY_MISSING],
+						'reinstallAppStoreApp'
+					);
+				}
 
-			// Try to update compatible apps
-			if($appsToUpgrade[self::KEY_COMPATIBLE]) {
-				$output->info('Attempting to update the following existing compatible apps from market: '.implode(', ', $appsToUpgrade[self::KEY_MISSING]));
-				$failedCompatibleApps = $this->getAppsFromMarket(
-					$output,
-					$appsToUpgrade[self::KEY_COMPATIBLE],
-					'upgradeAppStoreApp'
-				);
-			}
+				// Try to update compatible apps
+				if(!empty($appsToUpgrade[self::KEY_COMPATIBLE])) {
+					$output->info('Attempting to update the following existing compatible apps from market: '.implode(', ', $appsToUpgrade[self::KEY_MISSING]));
+					$failedCompatibleApps = $this->getAppsFromMarket(
+						$output,
+						$appsToUpgrade[self::KEY_COMPATIBLE],
+						'upgradeAppStoreApp'
+					);
+				}
 
-			$hasNotUpdatedCompatibleApps = count($failedCompatibleApps);
-		} catch (AppManagerException $e) {
-			$output->warning('No connection to marketplace: ' . $e->getPrevious());
+				$hasNotUpdatedCompatibleApps = count($failedCompatibleApps);
+			} catch (AppManagerException $e) {
+				$output->warning('No connection to marketplace: ' . $e->getPrevious());
+			}
+		} else {
+			// No market available, output error and continue attempt
+			$output->warning('Market app is unavailable for updating of apps. Enable with: occ:app:enable market');
 		}
 
 		$hasBlockingMissingApps = count($failedMissingApps);

--- a/tests/lib/Repair/AppsTest.php
+++ b/tests/lib/Repair/AppsTest.php
@@ -1,0 +1,71 @@
+<?php
+/**
+ * @author Tom Needham <tom@owncloud.com>
+ *
+ * @copyright Copyright (c) 2017, ownCloud GmbH.
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace Test\Repair;
+use OCP\App\IAppManager;
+use OCP\IConfig;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+/**
+ * Tests to check version comparison
+ *
+ * @see \OC\Repair\AppsTest
+ */
+class AppsTest extends \Test\TestCase {
+
+	/** @var \OC\Repair\AvatarPermissions */
+	protected $repair;
+
+	/** @var IAppManager */
+	protected $appManager;
+	/** @var  EventDispatcher */
+	protected $eventDispatcher;
+	/** @var  IConfig */
+	protected $config;
+
+	protected function setUp() {
+		parent::setUp();
+
+		$this->appManager = $this->getMockBuilder(IAppManager::class)->getMock();
+		$this->eventDispatcher = $this->getMockBuilder(EventDispatcher::class)->getMock();
+		$this->config = $this->getMockBuilder(IConfig::class)->getMock();
+		$this->repair = new \OC\Repair\Apps($this->appManager, $this->eventDispatcher, $this->config);
+	}
+
+	public function testMarketEnableVersionCompare10() {
+		$this->config->expects($this->once())->method('getSystemValue')->with('version', '0.0.0')->will($this->returnValue('10.0.0'));
+		$this->assertTrue($this->invokePrivate($this->repair, 'requiresMarketEnable'));
+	}
+
+	public function testMarketEnableVersionCompare9() {
+		$this->config->expects($this->once())->method('getSystemValue')->with('version', '0.0.0')->will($this->returnValue('9.1.5'));
+		$this->assertTrue($this->invokePrivate($this->repair, 'requiresMarketEnable'));
+	}
+
+	public function testMarketEnableVersionCompareFuture() {
+		$this->config->expects($this->once())->method('getSystemValue')->with('version', '0.0.0')->will($this->returnValue('10.0.2'));
+		$this->assertFalse($this->invokePrivate($this->repair, 'requiresMarketEnable'));
+	}
+
+	public function testMarketEnableVersionCompareCurrent() {
+		$this->config->expects($this->once())->method('getSystemValue')->with('version', '0.0.0')->will($this->returnValue('10.0.1'));
+		$this->assertFalse($this->invokePrivate($this->repair, 'requiresMarketEnable'));
+	}
+}

